Add SelectEvent to interaction

bjornharrtell opened the pull request:
This PR implements a coarse grained event directly on the Select interaction. The motive for this new event is that selection changes currently can only be observed on the collection of selected features and when selecting or deselecting many features many add/remove events will be dispatched which might be difficult to process as a unit in application code. This new event will only trigger once per interaction.

elemoine commented:
Anyone wants to review this for 3.2? I'll move it to v3.3.0 if not reviewed by tomorrow Wednesday.

@bjornharrtell I'd recommend squashing your commits into one.

bjornharrtell commented:
Squashed. Still looking to get this in so I would be greatful for any review.

elemoine commented:
If we make the select interaction trigger events shouldn't we introduce more specific events like `select` and `deselect`, with a `features` property indicating what features were selected or deselected. With your `change` event the user cannot know if a selection or a deselection occurred, nor can I know what features were selected or deselected.

bjornharrtell commented:
I agree that the information indicating what features were selected or deselected is desirable. I think this information should be added to `change` events rather than introducing separate events, to avoid the inconvenience of having to listen to two events.

elemoine commented:
You mean one event type with `selectedFeatures` and `deselectedFeatures` properties?

bjornharrtell commented:
Yes, but I suggest naming them simply `selected` and `deselected` (with documented type information).

bjornharrtell commented:
I've implemented properties as discussed and adjusted the select-features example to describe and use the event.

sweco-sebhar commented:
Rebased on master to resolve conflicts that was introduced by recent changes.

bjornharrtell commented:
Fixed build again and renamed event type to `select`.

elemoine commented:
If we add a `beforeselect` event type, don't you think we should have both `select` and `deselect`, and trigger `select`/`deselect` events for each selected/deselected feature?

acanimal commented:
Looking at this https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/ol3-dev/77k3E86dO8M/AvNUcWentzcJ it would be nice the select interaction returns, in addition to the feature, a reference to the layer the feature belongs to.

bjornharrtell commented:
@elemoine no I think it's more practical to only have to manage a single event subscription for selection/deselection.

@acanimal yes I have been missing a reference to the layer a few times when working with selection. However, I think it's possible (but uncommon?) for one feature to "belong" to multiple layers at the same time?

tschaub commented on examples/select-features.html:
```diff
@@ -35,6 +35,7 @@
           <h4 id="title">Select features example</h4>
           <p id="shortdesc">Example of using the Select interaction. Choose between <code>Single-click</code>, <code>Click</code> and <code>Hover</code> as the event type for selection in the combobox below. When using <code>Single-click</code> or <code>Click</code> you can hold do <code>Shift</code> key to toggle the feature in the selection.</p>
           <p>Note: when <code>Single-click</code> is used double-clicks won't select features. This in contrast to <code>Click</code>, where a double-click will both select the feature and zoom the map (because of the <code>DoubleClickZoom</code> interaction). Note that <code>Single-click</code> is less responsive than <code>Click</code> because of the delay it uses to detect double-clicks.</p>
+          <p>In this example, a listener is registered for the Select interaction's <code>change</code> event in order to update the selection status below.
```
It looks like this needs to be updated to read `select` instead of `change`.

bjornharrtell replied:
Thanks for noticing @tschaub, fixed now.

tschaub commented:
Since we've had to put out [two](https://github.com/openlayers/ol3/releases/tag/v3.1.1) [patch](https://github.com/openlayers/ol3/releases/tag/v3.2.1) releases in response to regressions in the select interaction behavior, I think we need to start building up some tests here.

elemoine commented:
> @acanimal yes I have been missing a reference to the layer a few times when working with selection.

+1

I'd like to be able to do that:

```js
selectInteraction.on('select',
    /**
     * @param {ol.interaction.SelectEvent} e Select event.
     */
    function(e) {
      var feature = e.feature;
      var layer = e.layer;
      // …
    });
```

But it's only possible if I get a `select` event for each selected feature, no?

> However, I think it's possible (but uncommon?) for one feature to "belong" to multiple layers at the same time?

If the same feature is drawn multiple times in multiple layers then, for that feature, you will get a `select` event for each layer the feature is drawn into.

sweco-sebhar commented:
My initial purpose with the select event in the first place was to support multiselect (see motivation in #3018). Without being able to think this though right now, a middle ground might perhaps be that select events are triggered per layer?

bartvde commented:
+1 on the need for tests here @tschaub

bartvde commented:
it seems https://github.com/bjornharrtell/ol3/commit/5d69007e289ad2fddf247fa403e346abf2c2ff62 got merged into this PR somehow @bjornharrtell ?

acanimal commented:
Hi @bjornharrtell 
maybe the elements stored in the `selected` and `deselected` array could be of a kind of type `selectedItem` composed of the feature, layers and any other future values (like clicked pixel and so on).

Currently I'm looking how to implement the `beforeselect` event. I'm not a master on Closure and I'm looking on the old `beforechangeproperty` code.

bjornharrtell commented:
Thanks for noticing @bartvde, I must have made an accidental merge. Should be fixed now.

@acanimal - yeah that is a good idea I think. It would solve the case @elemoine mentioned and allows the `select` to be triggered once per selection regardless of number of involved layers or features without adding too much complexity. +1 :)

bjornharrtell commented:
So I think the conclusion is to ignore `beforeselect` for now (see discussion in #3277). Let's keep this PR about the `select` event only which I still think would be very useful.

I've created basic test cases for the Select interaction in #3301 so I'm still hoping to get this merged eventually.

ahocevar commented on src/ol/interaction/selectinteraction.js:
```diff
@@ -13,6 +14,59 @@ goog.require('ol.interaction.Interaction');
+   * @type {ol.Collection.<ol.Feature>}
+   * @api
+   */
+  this.features = features;
```
I think this is redundant, since you can also use `ol.interaction.Select#getFeatures()`.

elemoine replied:
Agree. I wanted to make that same comment.

elemoine replied:
For example:

```js
selectInteraction.on('select', function(e) {
  var features = e.target.getFeatures();
  // …
});

bjornharrtell replied:
ol.SelectEvent#selected contains only the selected features that is the result of the current operation that triggered the event and might be subset of the features as returned by ol.interaction.Select#getFeatures(). This is used in the modified select-features example to present a total count togheter with last operation selected/deselected counts.

ahocevar commented on src/ol/interaction/selectinteraction.js:
```diff
@@ -156,7 +211,9 @@ ol.interaction.Select.handleEvent = function(mapBrowserEvent) {
     } else {
+      change = true;
       if (features.getLength() !== 0) {
+        Array.prototype.push.apply(deselected, features.getArray());
```
To avoid hitting the maximum function argument length, you should do
```js
deselected = [].concat(features.getArray());
```
Maybe not so elegant because it creates a new array, but it is safer.

elemoine replied:
Good point @ahocevar. I'd never thought about that.

ahocevar commented:
Thanks @bjornharrtell for the continued effort on this, and sorry for the back and forth. See two more minor comments above. If you are able to address them today, I can include this improvement in the 3.3.0 release.

elemoine commented on src/ol/interaction/selectinteraction.js:
```diff
@@ -13,6 +14,59 @@ goog.require('ol.interaction.Interaction');
+ * this type.
+ *
+ * @param {string} type The event type.
+ * @param {ol.Collection.<ol.Feature>} features Currently selected features
```
We usually end param/type/return description with a dot.

elemoine commented:
This also looks good to me. It'd be good to have this in 3.3.0.

sweco-sebhar commented:
Thanks! Fixed the jsdocs, array concat and rebased on master.

I didn't make any changes to l.SelectEvent#selected, see motivation in diff comments about that.

bjornharrtell commented:
Thanks! Fixed the jsdocs, array concat and rebased on master.

I didn't make any changes to l.SelectEvent#selected, see motivation in diff comments about that.

elemoine commented on src/ol/interaction/selectinteraction.js:
```diff
@@ -13,6 +14,59 @@ goog.require('ol.interaction.Interaction');
+   * @type {ol.Collection.<ol.Feature>}
+   * @api
+   */
+  this.features = features;
```
> ol.SelectEvent#selected contains only the selected features that is the result of the current operation that triggered the event and might be subset of the features as returned by ol.interaction.Select#getFeatures(). This is used in the modified select-features example to present a total count togheter with last operation selected/deselected counts.

Can't one get the collection using `evt.target.getFeatures()`?

elemoine commented on examples/select-features.js:
```diff
@@ -61,6 +61,11 @@ var changeInteraction = function() {
   if (select !== null) {
     map.addInteraction(select);
+    select.on('select', function(e) {
+      $('#status').html('&nbsp;' + e.features.getLength() +
```
Can't you use `e.target.getFeatures().getLength()` or `select.getFeatures().getLength()` here?

bjornharrtell commented:
Sorry I misunderstood and thought you where commenting ol.SelectEvent#selected and didn't consider ol.SelectEvent#features which is indeed redundant. Removed and fixed in example now.

sweco-sebhar commented:
In my latest squash it seems I must have forgot to comment out the squashed commit message so that "Fixed" was appended as a second line comment. Sorry.

ahocevar commented:
@sweco-sebhar I think we can live with that. Thanks for pointing it out.

fperucic commented:
What is the difference between  `ol.interaction.Select.getFeatures().on('add')` and `ol.interaction.Select.on('select')`?
I guess you get all the selected or deselected features at once when using the `select` event. Am I right?

acanimal commented:
Select.getFeatures() returns an ol.Collection, than means the on('add')
listens for any new object added to the collection.
The second form listens for a 'select' event in the select interaction.

Although the result can be similar you are doing (and listening) in
different objects.

Cheers.

On Tue, Mar 3, 2015 at 9:52 PM, Fran Peručić <notifications@github.com>
wrote:

> What is the difference between
> ol.interaction.Select.getFeatures().on('add') and
> ol.interaction.Select.on('select')?
> I guess you get all the selected or deselected features at once when using
> the select event. Am I right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openlayers/ol3/pull/3018#issuecomment-77032027>.
>



-- 
|----------------------------------------------------------------
| The book of OpenLayers 3 - http://kcy.me/x29y
| OpenLayers Cookbook: *http://kcy.me/ar2b <http://kcy.me/ar2b>*
| Blog: http://www.acuriousanimal.com/blog
| Twitter: @acanimal <http://twitter.com/#%21/acanimal>
|----------------------------------------------------------------
|   _      __
|  /_)    (_ ` _   _  _)_ o  _   _   _
| / /  o .__) (_( ) ) (_  ( (_( (_( (_)
|                                 _)
|----------------------------------------------------------------

bjornharrtell commented:
Yes, relevant for multiselecting in particular.

elemoine commented:
@fperucic, a `select` event is triggered as a result of a browser event, when the user selected and/or deselected features with the mouse/pointer/touch. `select` events are not triggered when features are programmatically selected/deselected, by adding/removing features to/from the select interaction's features collection. So it's a way to get notified when the user interacts with features.

tremby commented:
The example doesn't work. `e.selected` and `e.deselected` are `undefined`.

http://openlayers.org/en/v3.3.0/examples/select-features.html

bjornharrtell commented:
This is a known problem with release builds. Fixed in master with https://github.com/openlayers/ol3/pull/3324.
